src/generators/clientContextFileGenerator.ts

```diff
@@ -1,26 +1,35 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
-import { Project, PropertyDeclarationStructure } from "ts-morph";
+import {
+  Project,
+  PropertyDeclarationStructure,
+  ConstructorDeclaration,
+  CodeBlockWriter,
+  ClassDeclaration,
+  SourceFile
+} from "ts-morph";
 import { normalizeName, NameType } from "../utils/nameUtils";
 import { ClientDetails } from "../models/clientDetails";
 import { PackageDetails } from "../models/packageDetails";
 import { ParameterDetails } from "../models/parameterDetails";
-import {
-  ImplementationLocation,
-  SchemaType,
-  ConstantSchema
-} from "@azure-tools/codemodel";
+import { ImplementationLocation, SchemaType } from "@azure-tools/codemodel";
 import {
   getCredentialsCheck,
   getCredentialsParameter
 } from "./utils/parameterUtils";
+import { BaseUrlDetails } from "../transforms/urlTransforms";
 
 export function generateClientContext(
   clientDetails: ClientDetails,
   packageDetails: PackageDetails,
   project: Project
 ) {
+  const clientParams = clientDetails.parameters.filter(
+    param => param.implementationLocation === ImplementationLocation.Client
+  );
+  const requiredParams = getRequiredParameters(clientParams);
+  const hasCredentials = !!clientDetails.options.addCredentials;
```

Comment:
`--add-credentials` is an optional parameter because it adds a `TokenCredential` parameter to the generated `ServiceClient` and some generated clients don't want or need it.  However, I think the ones that don't want it are fewer than the ones that do, so maybe we should consider inverting the command line argument at some point

Reply:
We get these when users pass the `--add-credentials` flag to autorest i.e. `autorest ... --add-credentials`. 

APIs that don't have authentication (such as test server scenarios so far) don't need this.

Reply:
Does that mean those generated clients don't need any credentials at all? Or do they take a different kind that we don't know about and can't generate?

Reply:
I think so, although I think you can still pass Authentication headers through the options parameter to authenticate 
```ts
const client = new BodyComplexClient({customHeaders: {Authorization: "<AUTH_TOKEN>"}});
```

I may be wrong so please correct me if I'm wrong @daviwil